NETWORK_PROTOCOL.md

```diff
@@ -1 +1,102 @@
+
+Each message (frame) begins with a message type, which is a varint, followed by its contents. Messages are self-delimiting, because their structure is known from their type, and all fields are either of fixed size, self-delimited varints, or length-prefixed strings or byte arrays.
+
+## Protocol negotiation
+
+The protocol is versioned and supports future extensions. The client (i.e. the side that opened the connection) and the server do a loose handshake:
+
+    --> clientHello(version: varint, extensions: Array[Id])
+    <-- serverHello(version: varint, extensions: Array[Id])
+    
+This is a 'loose' handshake because the server doesn't have to wait for the `clientHello` before sending its `serverHello`. 
+    
+The protocol version is currently version 0. If either side receives a hello message with a version it doesn't support, it MUST send a `goodbye` message (defined below) and close the connection. When future versions of the protocol introduce incompatible changes and increment the version number, transports SHOULD indicate the incompatibility when suitable, e.g. by changing the HTTP Content-Type or TCP port number).
+    
+Extension to the protocol specify optional or future behaviors. 
+ 1. If an extension defines a new message type not described in this specification, that message MUST NOT be sent before receiving a hello from the other side confirming that it supports that extension. 
+ 2. If an extension changes the semantics of message types defined in this specification or by another extension, the modified behavior MUST be negotiated by at least one of the parties sending, and the other acknowledging, a message (defined by the extension being discussed) that declares the new behavior as active. A party supporting such an extension SHOULD NOT send messages whose semantics are modified by it before this negotiation is completed (i.e. the acknowledgement message is received).
+    
+The client can optimistically send more messages after the `clientHello` without waiting for the `serverHello`. If it eventually receieves a `serverHello` with a different protocol version, it must consider that its messages were discarded. Future protocol versions will not be backward-compatible with version 0, in the sense that if a server multiple versions (e.g. both version 0 and some future version 1), it must wait for the `clientHello` and then send a `serverHello` with a version number matching the client's.
+
+## The Reactive Streams core protocol
+
+The basic RS signalling is:
+
+    --> subscribe(publisher: String, subscriber: Id, initialDemand: Long = 0)
+    --> request(subscriber: Id, demand: Long)
+    --> cancel(subscriber: Id)
+    <-- subscribed(subscriber: Id, elementSize: varint = 0) // For elementSize != 0, see the next section
+    <-- onNext(subscriber: Id, element: bytes) 
+    <-- onComplete(subscriber: Id)
+    <-- onError(subscriber: Id, error: String)
+    
```

Comment:
How are you envisioning these signals being encoded over the wire? I'm assuming binary encoding would represent each of these signals as a byte rather than string of text? Is this the varint you reference in the message frame above?